cicecore/cicedynB/dynamics/ice_dyn_evp.F90

```diff
@@ -871,6 +1576,688 @@ subroutine stress (nx_block,   ny_block,   &
+         etax2T   , & ! 2*eta at the T point
+         shrU     , & ! shearU array
+         strength     ! ice strength at the T point
+      
+      real (kind=dbl_kind), dimension (nx_block,ny_block), intent(inout) :: &
+         stress12    ! sigma12
+
+      ! local variables
+
+      integer (kind=int_kind) :: &
+         i, j, ij
+
+      real (kind=dbl_kind), dimension (nx_block,ny_block) :: &
+         DeltaU       ! strain rates at U point
+
+      real (kind=dbl_kind) :: &
+         zetax2U, etax2U, rep_prsU,      & ! replacement pressure at U point
+         DminUarea, strtmp, areatmp        ! Dmin on U and tmp variables
+
+      character(len=*), parameter :: subname = '(stressC_U)'
+      
+      !-----------------------------------------------------------------
+      ! strain rates at U point
+      ! NOTE these are actually strain rates * area  (m^2/s)
+      !-----------------------------------------------------------------
+
+      if (visc_method == 'avg_strength') then
+         call strain_rates_U (nx_block     ,   ny_block    , &
+                              icellu       ,                 &
+                              indxui  (:)  , indxuj   (:)  , &
+                              uvelE   (:,:), vvelE    (:,:), &
+                              uvelN   (:,:), vvelN    (:,:), &
+                              uvelU   (:,:), vvelU    (:,:), &
+                              dxE     (:,:), dyN      (:,:), &
+                              dxU     (:,:), dyU      (:,:), &
+                              ratiodxN(:,:), ratiodxNr(:,:), &
+                              ratiodyE(:,:), ratiodyEr(:,:), &
+                              epm(:,:)     , npm(:,:)      , &
+                              DeltaU = DeltaU(:,:)           )
+      endif
+
+      do ij = 1, icellu
+         i = indxui(ij)
+         j = indxuj(ij)
+
+         !-----------------------------------------------------------------
+         ! viscosities and replacement pressure at U point
+         ! avg_zeta: Bouillon et al. 2013, C1 method of Kimmritz et al. 2016
+         ! avg_strength: C2 method of Kimmritz et al. 2016
+         !-----------------------------------------------------------------
+
+         if (visc_method == 'avg_zeta') then
+            DeltaU(i,j) = c0   ! not needed in avgzeta just computing etax2U
+            call visc_replpress_avgzeta (zetax2T (i  ,j  ), zetax2T (i  ,j+1), &
+                                         zetax2T (i+1,j+1), zetax2T (i+1,j  ), &
+                                         etax2T  (i  ,j  ), etax2T  (i  ,j+1), &
+                                         etax2T  (i+1,j+1), etax2T  (i+1,j  ), &
+                                         hm      (i  ,j  ), hm      (i  ,j+1), &
+                                         hm      (i+1,j+1), hm      (i+1,j  ), &
+                                         tarea   (i  ,j  ), tarea   (i  ,j+1), &
+                                         tarea   (i+1,j+1), tarea   (i+1,j  ), &
```

Comment:
Sending 'slices' of arrays like this can be expensive.  It looks like it would be more efficient to move the loop into the visc_replpress_* routines, i.e. calculate on entire arrays.  If you do that, then maybe you could just use the existing 4-point weighted averaging routines.

Reply:
I agree with this strategy in general.  It would require some additional arrays for the output variables, but think it could work out fine.  I'll take a look at that in greater detail and propose some changes.  I don't like calling these methods one gridcell at a time either.